content/blog/2025-03-20-new-hash-table.md

```diff
@@ -0,0 +1,202 @@
+
+We eliminated the dictEntry and instead embed key and value in the serverObject,
+along with other metadata for the key.
+
+```
++-----------+
+| hashtable |        bucket            bucket            bucket
++-----------+       +-----------------+-----------------+-----------------+-----
+| table 0  -------->| m x x x x x x x | m x x x x x x x | m x x x x x x x | ...
+| table 1   |       +-----------------+-----|-----------+-----------------+-----
++-----------+                               |
+                                            v
+                               +------------------------+
+                               | serverObject           |
+                               +------------------------+
+                               | type, encoding,        |
+                               | ref-counter, etc.      |
+                               | "FOO" (embedded key)   |
+                               | "BAR" (embedded value) |
+                               +------------------------+
+```
+
+Assuming the hashtable and the table are already in the CPU cache, looking up
+key-value entry now requires only two memory lookups: The bucket and the
+serverObject. If there is a hash collission, the object we're looking for is
+most likely in the same bucket, so no extra memory access is required.
+
+If a bucket becomes full, the last element slots in the bucket is replaced by a
+pointer to a child bucket. A child bucket has the same layout as a regular
+bucket, but it's a separate allocation. Child buckets form a chain. There is
+some probability of this happening, but long chains are very rare. Most of the
+keys are stored in top-level buckets.
+
+```
++-----------+
+| hashtable |        bucket            bucket            bucket
++-----------+       +-----------------+-----------------+-----------------+-----
+| table 0  -------->| m x x x x x x x | m x x x x x x c | m x x x x x x x | ...
+| table 1   |       +-----------------+---------------|-+-----------------+-----
++-----------+                                         |
+                                        Child bucket  v
+                                       +-----------------+
+                                       | m x x x x x x c |
+                                       +---------------|-+
+                                                       |
+                                         Child bucket  v
+                                        +-----------------+
+                                        | m x x x x x x x |
+                                        +-----------------+
+```
+
+Hashes, sets and sorted sets
+----------------------------
+
+The nested data types Hashes, Sets and Sorted sets also make use of the new hash
+table. The memory usage is reduced by roughly 10-20 bytes per entry. Memory and
+latency/throughput results are WIP.
+
+Iterator prefetching
+--------------------
```

Comment:
This is technically a separate point. It addresses some of the content of the blog, but ultimately feels a bit random at the end. Maybe we drop it and write another followup blog on memory prefetching?

Reply:
We could include it in the 8.1 overview blog post. I'm not sure anyone wants to write a prefetching blog post specifically.